cloudstate/eventsourced_servicer.py

```diff
@@ -0,0 +1,143 @@
+"""
+Copyright 2020 Lightbend Inc.
+Licensed under the Apache License, Version 2.0.
+"""
+
+import logging
+from pprint import pprint
+from typing import List
+
+from google.protobuf import symbol_database as _symbol_database
+
+from cloudstate.entity_pb2 import Command
+from cloudstate.event_sourced_context import (
+    EventContext,
+    EventSourcedCommandContext,
+    SnapshotContext,
+)
+from cloudstate.event_sourced_entity import EventSourcedEntity, EventSourcedHandler
+from cloudstate.event_sourced_pb2 import (
+    EventSourcedEvent,
+    EventSourcedInit,
+    EventSourcedReply,
+    EventSourcedSnapshot,
+    EventSourcedStreamOut,
+)
+from cloudstate.event_sourced_pb2_grpc import EventSourcedServicer
+from cloudstate.utils.payload_utils import get_payload, pack
+
+_sym_db = _symbol_database.Default()
+
+TYPE_URL_PREFIX = "type.googleapis.com/"
+
+
+class CloudStateEventSourcedServicer(EventSourcedServicer):
+    def __init__(self, event_sourced_entities: List[EventSourcedEntity]):
+        self.event_sourced_entities = {
+            entity.name(): entity for entity in event_sourced_entities
+        }
+
+    def handle(self, request_iterator, context):
+        initiated = False
+        current_state = None
+        handler: EventSourcedHandler = None
+        entity_id: str = None
+        start_sequence_number: int = 0
+        for request in request_iterator:
+            if not initiated:
+                if request.HasField("init"):
+                    init: EventSourcedInit = request.init
+                    service_name = init.service_name
+                    entity_id = init.entity_id
+                    if service_name not in self.event_sourced_entities:
+                        raise Exception(
+                            "No event sourced entity registered for service {}".format(
+                                service_name
+                            )
+                        )
+                    entity = self.event_sourced_entities[service_name]
+                    handler = EventSourcedHandler(entity)
+                    current_state = handler.init_state(entity_id)
+                    initiated = True
```

Comment:
This particular design detail isn't changed in this pr. The change is diffed because I moved the file: previously it was misnamed `evensourced_servicer`. Though I use a similar pattern in the stateless streaming handlers. It's an interesting question though. The `EventsourcedHandler` class is mostly just an ops adapter, and should have a very small constructor cost. The in-memory persistent objects, in this case, `self.event_sourced_entities` are the real underlying objects. The init flag is actually less to do with the object, and more to do with the stream handling mechanics.